src/dataflow/src/source/pubnub.rs

```diff
@@ -0,0 +1,61 @@
+use async_trait::async_trait;
+use futures::StreamExt;
+use pubnub_hyper::core::data::message::Type;
+use pubnub_hyper::{Builder, DefaultRuntime, DefaultTransport};
+
+use crate::source::{SimpleSource, SourceError, Timestamper};
+use dataflow_types::PubnubSourceConnector;
+use repr::{Datum, RowPacker};
+
+/// Information required to sync data from Pubnub
+pub struct PubnubSourceReader {
+    connector: PubnubSourceConnector,
+}
+
+impl PubnubSourceReader {
+    /// Constructs a new instance
+    pub fn new(connector: PubnubSourceConnector) -> Self {
+        Self { connector }
+    }
+}
+
+#[async_trait]
+impl SimpleSource for PubnubSourceReader {
+    async fn start(mut self, timestamper: &Timestamper) -> Result<(), SourceError> {
+        let transport = DefaultTransport::new()
+            // we don't need a publish key for subscribing
+            .publish_key("")
+            .subscribe_key(&self.connector.subscribe_key)
+            .build()
+            .map_err(SourceError::FileIO)?;
+
+        let mut pubnub = Builder::new()
+            .transport(transport)
+            .runtime(DefaultRuntime)
+            .build();
+
+        let channel = self.connector.channel.parse().unwrap();
+
+        let stream = pubnub.subscribe(channel).await;
```

Comment:
How does this handle connection failures mid stream? Can it resume from where it left off?

Reply:
It doesn't, I can try to add reconnect logic bug messages may be lost. That's just the nature of pubnub though so it's ok